app/controllers/profile.js

```diff
@@ -8,4 +10,285 @@ export default class ProfileController extends Controller {
+      required: false,
+      showError: false,
+      disabled: false,
+    },
+  ];
+
+  timerId = undefined;
+
+  //Debounce Function
+  debounce(func, delay) {
+    clearTimeout(this.timerId);
+    this.timerId = setTimeout(func, delay);
+  }
+
+  @action handleFieldChange(name, value) {
+    const index = this.fields.findIndex((field) => field.id === name);
+    set(this.formData, name, value);
+
+    if (this.fields[index].required && !value) {
+      this.isSubmitDisabled = true;
+      set(this.fields[index], 'showError', true);
+    } else if (!this.fields[index].validator) {
+      set(this.fields[index], 'showError', false);
+    }
+
+    const anyErrors = this.fields.map((field) => {
+      if (field.required && this.formData[field.id] === '') {
+        return true;
+      }
+      return false;
+    });
+
+    if (anyErrors.filter(Boolean).length) {
+      this.isSubmitDisabled = true;
+    } else {
+      this.isSubmitDisabled = false;
+    }
+  }
+
+  async checkUserName(userName) {
+    if (!userName) {
+      return set(this.fields[2], 'errorMessage', 'Username cannot be empty');
+    }
+    try {
+      const lowerCaseUsername = userName.toLowerCase();
+      const response = await fetch(
+        `${BASE_URL}/users/isUsernameAvailable/${lowerCaseUsername}`,
+        {
+          method: 'GET',
+          headers: {
+            'Content-Type': 'application/json',
+          },
+          credentials: 'include',
+        }
+      );
+      const data = await response.json();
+      const { isUsernameAvailable } = data;
+      set(this.fields[2], 'showError', !isUsernameAvailable);
+      set(
+        this.fields[2],
+        'errorMessage',
+        `${userName} is not available, Choose another username!`
+      );
+    } catch (error) {
+      console.error('Error : ', error);
+    }
+  }
+
+  @action async userNameValidator(userName) {
+    this.debounce(() => this.checkUserName(userName), 500);
+  }
+
+  @action phoneNumberValidator(phone) {
+    if (typeof phone !== 'string') {
+      return false;
+    }
+
+    const pattern = /^(0|[+91]{3})?[7-9][0-9]{9}$/;
+    const index = this.fields.findIndex((field) => field.id === 'phone');
+
+    if (pattern.test(phone)) {
+      set(this.fields[index], 'showError', false);
+    } else {
+      set(this.fields[index], 'showError', true);
+    }
+  }
+
+  @action emailValidator(email) {
+    if (typeof email !== 'string') return false;
+
+    const pattern = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;
+    const index = this.fields.findIndex((field) => field.type === 'email');
+
+    if (pattern.test(email)) {
+      set(this.fields[index], 'showError', false);
+    } else {
+      set(this.fields[index], 'showError', true);
+    }
+    return;
+  }
+
+  removeEmptyFields(reqObject) {
+    for (const field in reqObject) {
+      if (!reqObject[field]) {
+        delete reqObject[field];
+      }
+    }
+    return reqObject;
+  }
```

Comment:
Directly updating the object here is not good. Why not destructure `reqObject` in a newer object and then perform iteration and return it? 🤔